src/inmanta/server/services/orchestrationservice.py

```diff
@@ -63,277 +63,204 @@
         """
-        rids_in_partial_update = {resource_minimal.get_resource_id_str() for resource_minimal in self.partial_updates}
-        rids_not_in_partial_compile = {rid for rid in merged_resources if rid not in rids_in_partial_update}
         old_unresolved_unknowns_to_keep = [
-            uk.copy(self.new_version)
-            for uk in await data.UnknownParameter.get_list(environment=self.env.id, version=self.base_version, resolved=False)
-            # Always keep unknowns not tied to a specific resource
-            if not uk.resource_id or uk.resource_id in rids_not_in_partial_compile
+            uk.copy(self.version)
+            for uk in await data.UnknownParameter.get_unknowns_to_copy_in_partial_compile(
+                environment=self.env_id,
+                source_version=self.base_version,
+                updated_resource_sets=self.updated_resource_sets,
+                deleted_resource_sets=self.deleted_resource_sets,
+                rids_in_partial_compile=self.rids_in_partial_compile,
+            )
         ]
-        return [*old_unresolved_unknowns_to_keep, *self.unknowns]
-
-    async def apply_partial(self, *, connection: asyncpg.connection.Connection) -> MergedModel:
-        """
-        Applies the partial model's resources on this instance's base version. The caller should acquire appropriate locks on
-        the database connection as defined in the put_partial method definition.
-
-        :param connection: The database connection to use to determine the latest version. Appropriate locks are assumed to be
-            acquired.
-        :return: A tuple of the resources and the resource sets. All resource version ids are set to this instance's new
-            version.
-        """
-        old_resources: dict[ResourceIdStr, ResourceWithResourceSet] = await self._get_base_resources(connection=connection)
-        old_resource_sets: Dict[ResourceIdStr, Optional[str]] = {
-            res_id: res.resource_set for res_id, res in old_resources.items()
-        }
-        paired_resources_partial = self.pair_resources_partial_update_to_old_version(old_resources)
-        updated_resource_sets: Set[str] = set(
-            res.new_resource.resource_set for res in paired_resources_partial if not res.new_resource.is_shared_resource()
-        )
-
-        merged_resources: dict[ResourceIdStr, ResourceWithResourceSet] = self._merge_resources(
-            old_resources, paired_resources_partial, updated_resource_sets
-        )
-        return MergedModel(
-            resources=[r.resource.dict() for r in merged_resources.values()],
-            resource_states={rid: r.get_resource_state_for_new_resource() for rid, r in merged_resources.items()},
-            resource_sets=self._merge_resource_sets(old_resource_sets, updated_resource_sets),
-            unknowns=await self._merge_unknowns(merged_resources),
-        )
+        return [*old_unresolved_unknowns_to_keep, *unknowns_in_partial_compile]
```

Comment:
Why do we treat these differently from the resources? The resources we just copy in `_put_version`, but these we still take through Python. Is there a reason behind that?

Reply:
I mainly didn't do this because the `UnknownParameter` table has id (uuid) as primary key. If we would copy the records from the base version to the new version, we would have to generate a new id from each copied unknown. Postgresql seems to have a [function](https://www.postgresql.org/docs/13/functions-uuid.html) for that, but it's only available in PG13+ and not in PG10. As the number of unknowns are usually small, I thought it wouldn't be an issue to do the merging in Python. On the other hand we merge this change only into ISO6+ which requires PG13, so it wouldn't be much of an issue. But it might break some of our internal labs.

Reply:
Your reasoning makes sense to me, you can consider this conversation resolved.

As to your final comment, I'm not sure how much of our infra still runs PG<13. @wouterdb, you added support for PG13 in the postgresql module in inmanta/postgresql#239. Did you also update the test orchestrators and other clients to use PG13, or is that still outstanding?

Reply:
We still support RHEL7, so I think we need to stay compatible with older PG's for a while.

Reply:
I don't think we do support RHEL7 for iso5+ though.

EDIT: man, it's annoying that GitHub doesn't allow you to comment directly while having a review in progress...

Reply:
See [iso4 docs](https://docs.inmanta.com/inmanta-service-orchestrator/4/install/1-install-server.html#step-2-install-postgresql-10) vs [iso5 docs](https://docs.inmanta.com/inmanta-service-orchestrator/5/install/1-install-server.html#step-2-install-postgresql-13). Both RHEL versions and Postgres version are mentioned there.